.github/workflows/ci-sage.yml

```diff
@@ -0,0 +1,173 @@
+##    as a build artifact named upstream.
+##
+##  - Second, it checks out a copy of the SageMath source tree.
+##    It downloads the upstream artifact and replaces the project's
+##    package in the SageMath distribution by the newly packaged one
+##    from the upstream artifact, by running the script "update-pkgs.sh".
+##    Then it builds a small portion of the Sage distribution.
+##
+## Many copies of the second step are run in parallel for each of the tested
+## systems/configurations.
+
+on:
+  pull_request:
+    types: [opened, synchronize]
+  push:
+    tags:
+      - '*'
+  workflow_dispatch:
+    # Allow to run manually
+
+env:
+  # Ubuntu packages to install so that the project's "make dist" can succeed
+  DIST_PREREQ:      python3-setuptools
+  # Name of this project in the Sage distribution
+  SPKG:             sympy
+  # Sage distribution packages to build
+  TARGETS_PRE:      build/make/Makefile
+  TARGETS:          SAGE_CHECK=no SAGE_CHECK_PACKAGES=sympy build ptest
+  TARGETS_OPTIONAL: build/make/Makefile
+  # Standard setting: Test the current beta release of Sage:
+  SAGE_REPO:        sagemath/sage
+  SAGE_REF:         develop
+  # Uncomment to temporarily test with the branch from a sage ticket
+  # (this is a no-op after that ticket is merged)
+  ## SAGE_TRAC_GIT:    https://github.com/sagemath/sagetrac-mirror.git
+  ## SAGE_TICKET:
+  ## REMOVE_PATCHES:   "*"
+
+jobs:
+
+  dist:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Check out ${{ env.SPKG }}
+        uses: actions/checkout@v2
+        with:
+          path: build/pkgs/${{ env.SPKG }}/src
+      - name: Install prerequisites
+        run: |
+          sudo DEBIAN_FRONTEND=noninteractive apt-get update
+          sudo DEBIAN_FRONTEND=noninteractive apt-get install $DIST_PREREQ
+      - name: Run make dist, prepare upstream artifact
+        run: |
+          (cd build/pkgs/${{ env.SPKG }}/src && python3 setup.py sdist ) \
+          && mkdir -p upstream && cp build/pkgs/${{ env.SPKG }}/src/dist/*.tar.gz upstream/${{ env.SPKG }}-git.tar.gz \
+          && echo "sage-package create ${{ env.SPKG }} --version git --tarball ${{ env.SPKG }}-git.tar.gz --type=standard" > upstream/update-pkgs.sh \
+          && if [ -n "${{ env.REMOVE_PATCHES }}" ]; then echo "(cd ../build/pkgs/${{ env.SPKG }}/patches && rm -f ${{ env.REMOVE_PATCHES }}; :)" >> upstream/update-pkgs.sh; fi \
+          && ls -l upstream/
+      - uses: actions/upload-artifact@v2
+        with:
+          path: upstream
+          name: upstream
+
+  docker:
+    runs-on: ubuntu-latest
+    needs: [dist]
+    strategy:
+      fail-fast: false
+      max-parallel: 32
+      matrix:
+        # debian-buster uses system python 3.7
+        # ubuntu-groovy uses system python 3.8
+        # fedora-34 uses system python 3.9
+        # archlinux-latest is at the cutting edge
+        tox_system_factor: [debian-buster, ubuntu-groovy, fedora-34, archlinux-latest]
```

Comment:
This seems like a fragile way to test various versions of Python. Whenever a new version of Python is released we have to find a new linux distro that includes it. Is it not possible to just build a docker container specifically for Sage with all the Python versions required? Or alternately use conda environments, since all the Sage dependencies are on conda-forge (assuming they work with the development version).

Reply:
It's also opaque in the test matrix. We should at least name the matrix entries appropriately, assuming that's possible.

Reply:
Actually, is it really necessary for us to test every version of Python? Do we anticipate problems that only occur in one version?

Reply:
The comments there just explain my selection among the dozens of platforms that we test in Sage.

I am not suggesting that you should replace your testing of supported Python versions by my CI.

Reply:
Well, this proposed workflow does integration testing with Sage (= the Sage distribution), which is what is most relevant for Sage development.

Using conda-forge would certainly be another way for sympy to test integration with some version of sage; but you won't get testing with the Sage development releases that way.

Reply:
> Actually, is it really necessary for us to test every version of Python? Do we anticipate problems that only occur in one version?

There are many differences between these 4 platforms, not just the Python version. Sage makes use of system packages that its `configure` scripts determines to be suitable. For example, it picks up openblas from the system; so if sympy contains numerical tests using numpy, the workflow will test it with that on these platforms. Whether this is useful for you, you will have to decide. It's fine with me if you want to test fewer platforms than the ones I put there.